tools/clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -1791,11 +1792,31 @@ void CodeGenFunction::EmitStoreThroughExtVectorComponentLValue(RValue Src,
       }
       Builder.CreateStore(Vec, VecDstPtr);
     } else {
-      for (unsigned i = 0; i < VecTy->getVectorNumElements(); i++) {
-        if (llvm::Constant *Elt = Elts->getAggregateElement(i)) {
-          llvm::Value *EltGEP = Builder.CreateGEP(VecDstPtr, {Zero, Elt});
-          llvm::Value *SrcElt = Builder.CreateExtractElement(SrcVal, i);
-          Builder.CreateStore(SrcElt, EltGEP);
+      // If the vector pointer comes from a call inst, assume it's a subscript.
+      // Do a load and then a store to avoid generating multiple stores.
+      if (llvm::CallInst *CI = llvm::dyn_cast<llvm::CallInst>(VecDstPtr)) {
```

Comment:
This seems like a non-obvious assumption. Not real familiar with the code here so maybe it is valid? Is there any stronger check we can do here to detect when it is actually a subscript?

Reply:
Added a different check. Also added check that the resource is a typed buffer.

Although if we do this change anyway for things like RW structured buffer, the RW structured buffer goes from having 3 stores (1 per component) to have 1 load and 1 store.